neo/rawio/intanrawio.py

```diff
@@ -183,28 +285,57 @@ def _get_analogsignal_chunk(self, block_index, seg_index, i_start, i_stop, strea
         mask = self.header["signal_channels"]["stream_id"] == stream_id
         signal_channels = self.header["signal_channels"][mask]
+
+        # for channel slicing need to keep track of if None or an actual array that can be indexed
+        channel_indexes_are_none = False
         if channel_indexes is None:
             channel_indexes = slice(None)
+            channel_indexes_are_none = True
         channel_ids = signal_channels["id"][channel_indexes]
 
-        shape = self._raw_data[channel_ids[0]].shape
+        if self.file_format == 'header-attached':
+            shape = self._raw_data[channel_ids[0]].shape
+        elif self.file_format == 'one-file-per-signal':
+            shape = self._raw_data[stream_index].shape
+        else:
+            if channel_indexes_are_none:
+                shape = self._raw_data[stream_index][0].shape
+            else:
+                shape = self._raw_data[stream_index][channel_indexes[0]].shape
 
         # some channel (temperature) have 1D field so shape 1D
         # because 1 sample per block
         if len(shape) == 2:
-            # this is the general case with 2D
+            # this is the general case with 2D (ie in the header-attached format)
             block_size = shape[1]
             block_start = i_start // block_size
             block_stop = i_stop // block_size + 1
 
             sl0 = i_start % block_size
             sl1 = sl0 + (i_stop - i_start)
 
-        sigs_chunk = np.zeros((i_stop - i_start, len(channel_ids)), dtype="uint16")
+        if self.file_format != 'header-attached' and stream_id == 'RHD2000 amplifier channel':
+            sigs_chunk = np.zeros((i_stop - i_start, len(channel_ids)), dtype="int16")
+        else:
+            sigs_chunk = np.zeros((i_stop - i_start, len(channel_ids)), dtype="uint16")
         for channel_index, channel_id in enumerate(channel_ids):
+            if self.file_format == 'header-attached':
             # Memmap fields are the channel_ids for unique channels
-            data_chan = self._raw_data[channel_id]
-            if len(shape) == 1:
+                data_chan = self._raw_data[channel_id]
+            elif self.file_format == 'one-file-per-signal':
+                if channel_indexes_are_none:
+                    data_chan = self._raw_data[stream_index][:, channel_index]
+                else:
+                    data_chan = self._raw_data[stream_index][:, channel_indexes[channel_index]]
+            else:
+                if channel_indexes_are_none:
+                    data_chan = self._raw_data[stream_index][channel_index]
+                else:
+                    data_chan = self._raw_data[stream_index][channel_indexes[channel_index]]
+
+            # if shape is 1 it is either temp a 1d or a 1d array from one-file-per-stream
+            # or one-file-per-channel. In which case we do not want to do the block math
+            if len(shape) == 1 or self.file_format != 'header-attached':
```

Comment:
@h-mayorquin I just realized we (but really my fault since I wrote this :( ) had another tiny bug that may mess things up. We shouldn't get the signal chunk based on the block_start and block_stop because for the non-header-attached formats we don't need to do this . We can just grab the starting index and the stopping index.

Reply:
And now I realize this comment is cutting off my point haha. If you look from line 336-345ish you'll see what I mean.

Reply:
Yep. That makes sense.

Reply:
Reading this function again I would suggest completly separating the workflow for `_get_analogsignal_chunk` for the three formats:

1) The header-attached version needs to block logic because the data is interleaved but is not necessary for the other formats.
2) The one-file-per-signal version does not need the loop over the channels as all the channels are in the memmap already and I think we can optimize access by using this layout.
3) The one-file-per-channel version requires iterating over channels for accessing the different memmaps but does not require block logic.

I personally would create three private methods:
1) `_get_analogsignal_chunk_header_attached`, `_get_analogsignal_chunk_one_file_per_signal`, `_get_analogsignal_chunk_one_file_per_channel` and just call them after the channel_index logic depending on the format.

You can also just three big if if you don't like adding methods. 

But I think this would be better for debugging, optimizing and profiling. I think that the fact that the other two modes do not require the block logic escape both of us because the current implementation is a little bit too complicated with so many branches. 

What do you think?

Reply:
I agree with that. Let's merge #1416 and then I'll draft that change while incorporating the typecasting changes at the same time.

Reply:
Do you want me to do this? I have been making a bunch of request. I can open a PR with this as a base and you can review it and merge to this branch if you aprove.

Reply:
Yes, just push your latest changes if you have anything to the remote of this branch and I will take care of it. I will also propagate the merge of the dtype stuff so you don't have to take care of it.

Reply:
I have no remote you're good to go. And #1416 is merged.

Reply:
Thanks Heberto!